fairlearn/reductions/exponentiated_gradient/exponentiated_gradient.py

```diff
@@ -6,256 +6,116 @@
-            + self.constraints.signed_weights(lambda_vec)
-        redY = 1 * (signed_weights > 0)
-        redW = signed_weights.abs()
-        redW = self.n * redW / redW.sum()
-
-        classifier = pickle.loads(self.pickled_learner)
-        classifier.fit(self.X, redY, redW)
-        self.n_oracle_calls += 1
-
-        def h(X): return classifier.predict(X)
-        h_error = self.obj.gamma(h)[0]
-        h_gamma = self.constraints.gamma(h)
-        h_value = h_error + h_gamma.dot(lambda_vec)
-
-        if not self.hs.empty:
-            values = self.errors + self.gammas.transpose().dot(lambda_vec)
-            best_idx = values.idxmin()
-            best_value = values[best_idx]
-        else:
-            best_idx = -1
-            best_value = np.PINF
-
-        if h_value < best_value - _PRECISION:
-            if self.debug:
-                print("%sbest_h: val improvement %f" %
-                      (_LINE, best_value - h_value))
-            h_idx = len(self.hs)
-            self.hs.at[h_idx] = h
-            self.classifiers.at[h_idx] = classifier
-            self.errors.at[h_idx] = h_error
-            self.gammas[h_idx] = h_gamma
-            best_idx = h_idx
-
-        return self.hs[best_idx], best_idx
-
-
-def _mean_pred(dataX, hs, weights):
-    """Return a weighted average of predictions produced by
-    classifiers in hs"""
-    pred = pd.DataFrame()
-    for t in range(len(hs)):
-        pred[t] = hs[t](dataX)
-    return pred[weights.index].dot(weights)
+from fairlearn.reductions import moments
+from fairlearn.reductions import ReductionsEstimator
+from ._constants import _ACCURACY_MUL, _REGRET_CHECK_START_T, _REGRET_CHECK_INCREASE_T, \
+    _SHRINK_REGRET, _SHRINK_ETA, _MIN_T, _RUN_LP_STEP, _PRECISION, _INDENTATION
+from ._lagrangian import _Lagrangian
 
+logger = logging.getLogger(__name__)
 
-# Explicit optimization parameters of expgrad
 
-# A multiplier controlling the automatic setting of nu.
-_ACCURACY_MUL = 0.5
-
-# Parameters controlling adaptive shrinking of the learning rate.
-_REGRET_CHECK_START_T = 5
-_REGRET_CHECK_INCREASE_T = 1.6
-_SHRINK_REGRET = 0.8
-_SHRINK_ETA = 0.8
-
-# The smallest number of iterations after which expgrad terminates.
-_MIN_T = 5
-
-# If _RUN_LP_STEP is set to True, then each step of exponentiated
-# gradient is followed by the saddle point optimization over the convex
-# hull of classifiers returned so far.
-_RUN_LP_STEP = True
-
-column_names = str("best_classifier best_gap classifiers "
-                   "weights last_t best_t n_oracle_calls")
-expgrad_result = namedtuple("ExpgradResult",
-                            column_names)
+def _mean_pred(X, hs, weights):
+    """Return a weighted average of predictions produced by classifiers in hs"""
+    pred = pd.DataFrame()
+    for t in range(len(hs)):
+        pred[t] = hs[t](X)
+    return pred[weights.index].dot(weights)
 
 
-def exponentiated_gradient(dataX, dataA, dataY, learner,
+class ExponentiatedGradientResult:
```

Comment:
Why are we doing this instead of using the named tuples?

Reply:
We had a thread about that in this PR. It's always a trade-off, do you want structure and extensibility, or something simple that "just works". namedtuple is cute and nice for a couple, or maybe three or four things, but here we had six or seven, so that becomes tricky and less readable. Richard's opinion was that we use well defined objects elsewhere so this should probably be a real class, too, and I agreed.

Reply:
It was actually advice from Ilya and Walter. I know that the difference between a class and named tuple in Python is minimal.